### DISCUSSION TITLE
TimeDisplay problem (maybe?)

### DISCUSSION THREAD

**Discussion by Vahen1981** [tttapa/Control-Surface#563] [Q&A]
Hi Pieter,
Since two weeks I started testing my control surface working in some music productions with my daw. I realized that after half an hour of working my DAW starts to goes slow... after two hours it is almost impossible to work, and everything fixes when I turn off my control surface.
I comment some lines in my code and test it again and I discover that my problem is time display. I am using a teensy 4.0 and a LCD 1602 through I2C to display the time. I suspect the problem is the I2C, what do you think? maybe is my code, but I don't think so, anyway here are the lines that I wrote for time display:

```
#include <LiquidCrystal_I2C_Wire2.h>
#include <Wire.h>
#include <Control_Surface.h>


LiquidCrystal_I2C lcd(0x27,16,2);


USBMIDI_Interface midiusb;


MCU::TimeDisplay timedisplay = {{CHANNEL_16}};

void showTime() {
  char tics[4];
  timedisplay.getText(tics, 7, 3);
  char frames[3];
  timedisplay.getText(frames, 5, 2);
  char beats[3];
  timedisplay.getText(beats, 3, 2);
  char bars[4];
  timedisplay.getText(bars, 0, 3);
  char allTime[14] = "";
  strcat(allTime, bars);
  strcat(allTime, ".");
  strcat(allTime, beats);
  strcat(allTime, ".");
  strcat(allTime, frames);
  strcat(allTime, ".");
  strcat(allTime, tics);
  Wire2.beginTransmission(0x27);
  lcd.setCursor(3,0);
  lcd.print(allTime);
  lcd.setCursor(5,1);
  lcd.print("SMPTE/Beats");
  Wire2.endTransmission();
}


void checkTime(){
  if(timedisplay.getDirty()){
    timedisplay.clearDirty();
    showTime();
  }
  else {}
}

void setup() {
  Control_Surface.begin();
  lcd.init();                      
  lcd.backlight();
  Wire.begin();
  Wire.setClock(400000);
  timedisplay.clearDirty();
}

void loop() {
    Control_Surface.loop();
    checkTime();
}
```

Please help me with your impressions!

**Comment by tttapa** [comment]
An I2C display is relatively slow, I'd recommend only updating it every 50 ms or so.

While the I2C transmission is in progress, the Teensy cannot handle incoming MIDI messages, so the DAW cannot send any new messages, and has to wait, causing the whole interface to lag.

**Reply by Vahen1981** [reply]
Thanks! I suspect something like that, but I didn't know how to solve it. To update every 50ms sounds a good idea, I'll try that, thanks again.

**Reply by Veroledez** [reply]
Hola, podrias  algo asi:
unsigned long tiempo =0;

En el void loop agrega

if ((millis()- tiempo) > 100 && timedisplay.geDirty()) {
showtime();
timedisplay.clearDirty();
tiempo =milli();
}
else {timedisplay.learDirty(); tiempo= millis();}

**Reply by tttapa** [reply]
I'd probably use something like this:
```cpp
Timer<millis> displayTimer {50};
void checkTime(){
  if(displayTimer && timedisplay.getDirty()) {
    timedisplay.clearDirty();
    showTime();
  }
}
```
You don't want to clear the dirty flag unless you've updated the display.

**Reply by Veroledez** [reply]
Sin duda, eso es mas elegante que lo que he propuesto, lo empezaré a usar . Un saludo

**Reply by Vahen1981** [reply]
Thanks Pieter, gracias Veroledez... I fixed it and now my daw works fast!
